File: packages/gbnf-experiment/data/20260909T024849Z_b96078c6/ported_implementation/src/rules-builder/symbol-ids.ts
/**
 * A `Map` of rule name -> rule id, that additionally remembers the position in the
 * grammar each symbol was declared at, and supports looking a name up by its id.
 */
export class SymbolIds extends Map<string, number> {
  // Lazily created: `Map`'s constructor may call `set` before field initialisers run.
  declare positions: Map<string, number> | undefined;
  declare reverseMapping: Map<number, string> | undefined;

  set(key: string, value: number, pos = 0): this {
    super.set(key, value);
    (this.reverseMapping ??= new Map()).set(value, key);
    (this.positions ??= new Map()).set(key, pos);
    return this;
  }

  /** Alias of `entries()`, mirroring the reference implementation. */
  items(): IterableIterator<[string, number]> {
    return this.entries();
  }

  reverseGet(key: number): string {
    const val = this.reverseMapping?.get(key);
    if (val === undefined) {
      throw new Error(`SymbolIds does not contain value: ${key}`);
    }
    return val;
  }

  getPos(key: string): number {
    const val = this.positions?.get(key);
    if (val === undefined) {
      throw new Error(`SymbolIds does not contain key: ${key}`);
    }
    return val;
  }

  toJSON(): Record<string, number> {
    return Object.fromEntries(this.entries());
  }
}

File: proposed-projects/round-trip-experiment/reverse/20260909T004354Z_a47a3057/20260910T044103Z_5267f914/ported_implementation/gbnf/rules_builder/symbol_ids.py
from typing import Dict, Generator, ItemsView, Optional, Tuple


class SymbolIds:
    def __init__(self) -> None:
        self._map: Dict[str, int] = {}
        self._pos: Dict[str, int] = {}
        self._reverse_map: Dict[int, str] = {}

    def entries(self) -> ItemsView[str, int]:
        return self._map.items()

    def __iter__(self) -> Generator[Tuple[str, int], None, None]:
        yield from list(self._map.items())

    @property
    def size(self) -> int:
        return len(self._map)

    def get(self, key: str) -> Optional[int]:
        return self._map.get(key)

    def set(self, key: str, value: int, pos: int) -> None:
        self._map[key] = value
        self._reverse_map[value] = key
        self._pos[key] = pos

    def has(self, key: str) -> bool:
        return key in self._map

    def reverse_get(self, key: int) -> str:
        val = self._reverse_map.get(key)
        if val is None:
            raise ValueError(f"SymbolIds does not contain value: {key}")
        return val

    def get_pos(self, key: str) -> int:
        val = self._pos.get(key)
        if val is None:
            raise ValueError(f"SymbolIds does not contain key: {key}")
        return val
